File: javascript/test/transaction-integrity-test.ts
import { LaserstreamClient, CommitmentLevel } from '../index';
import { SubscribeUpdate } from '@triton-one/yellowstone-grpc';
import Client from '@triton-one/yellowstone-grpc';
import bs58 from 'bs58';

// Load test configuration
const testConfig = require('../test-config.js');

async function main() {
  const PUMP = 'pAMMBay6oceH9fJKBRHGP5D4bD4sWpmSwMn52FMfXEA';

  // Laserstream stream (under test)
  const client = new LaserstreamClient(testConfig.laserstream.endpoint, testConfig.laserstream.apiKey);


  const subscriptionRequest: any = {
    transactions: {
      client: {
        accountInclude: [PUMP],
        accountExclude: [],
        accountRequired: [],
        vote: false,
        failed: false
      }
    },
    commitment: CommitmentLevel.Confirmed,
    accounts: {},
    slots: {},
    transactionsStatus: {},
    blocks: {},
    blocksMeta: {},
    entry: {},
    accountsDataSlice: [],
  };

  // Yellowstone node for comparing with Laserstream
  const yellowstoneConfig = {
    endpoint: testConfig.yellowstone.endpoint,
    xToken: testConfig.yellowstone.apiKey
  } as const;


  const gotLS = new Set<string>(); // laserstream signatures
  const gotYS = new Set<string>(); // yellowstone signatures

  // Store latest slot for each signature so we can give more context when reporting mismatches.
  const slotLS = new Map<string, string>();
  const slotYS = new Map<string, string>();
  let newLS = 0;
  let newYS = 0;
  let errLS = 0;
  let errYS = 0;

  // --- per-slot tracking to avoid premature mismatch logging ---
  const lsBySlot = new Map<number, Set<string>>();
  const ysBySlot = new Map<number, Set<string>>();
  let maxSlotLS = 0;
  let maxSlotYS = 0;
  const SLOT_LAG = 3000; // number of slots to wait before declaring a slot closed

  const INTEGRITY_CHECK_INTERVAL_MS = 30_000; // 30 seconds
  setInterval(async () => {
    const readySlot = Math.min(maxSlotLS, maxSlotYS) - SLOT_LAG; // wait additional slots for late txs

    const slotsToProcess = new Set<number>();
    for (const s of lsBySlot.keys()) if (s <= readySlot) slotsToProcess.add(s);
    for (const s of ysBySlot.keys()) if (s <= readySlot) slotsToProcess.add(s);

    let totalMissingLS = 0;
    let totalMissingYS = 0;

    for (const slot of slotsToProcess) {
      const setLS = lsBySlot.get(slot) ?? new Set<string>();
      const setYS = ysBySlot.get(slot) ?? new Set<string>();

      const missingLS = new Set<string>(); // present in YS but not LS
      const missingYS = new Set<string>(); // present in LS but not YS

      for (const sig of setYS) if (!setLS.has(sig)) missingLS.add(sig);
      for (const sig of setLS) if (!setYS.has(sig)) missingYS.add(sig);

      if (missingLS.size || missingYS.size) {
        console.error(`[INTEGRITY] transaction_mismatch slot=${slot}  missing Laserstream=${missingLS.size} missing Yellowstone=${missingYS.size}`);
        for (const sig of missingLS) console.error(`SIGNATURE MISSING IN LASERSTREAM ${sig}  (YS_slot=${slot})`);
        for (const sig of missingYS) console.error(`SIGNATURE MISSING IN YELLOWSTONE ${sig}  (LS_slot=${slot})`);
      }

      totalMissingLS += missingLS.size;
      totalMissingYS += missingYS.size;

      lsBySlot.delete(slot);
      ysBySlot.delete(slot);
    }

    const now = new Date().toISOString();
    console.log(`[${now}] laserstream+${newLS}  yellowstone+${newYS}  processedSlots:${slotsToProcess.size} missingLS:${totalMissingLS} missingYS:${totalMissingYS}  LS_errors:${errLS}  YS_errors:${errYS}`);

    // reset counters (keep global sets for history but clear new counters)
    newLS = 0;
    newYS = 0;
  }, INTEGRITY_CHECK_INTERVAL_MS);

  // Transaction integrity test started

  // Helper: narrow update to transaction variant – we work with `any` here to avoid heavy protobuf typings(idk how to do this better)
  function isTransactionUpdate(update: any): update is { transaction: any } {
    return typeof (update as any).transaction === 'object' && (update as any).transaction !== null;
  }

  // Helper: reliably extract a base-58 signature string from a transaction update
  function extractSigAndSlot(update: any): {sig: string | null; slot: string | null} {
    if (!isTransactionUpdate(update)) return {sig: null, slot: null} as any;

    const info: any | undefined = (update.transaction as any)?.transaction;

    if (!info) return {sig: null, slot: null} as any;

    const rawSig: Uint8Array | undefined = (info.signature ?? info.transaction?.signature) as Uint8Array | undefined;
    if (!rawSig) return {sig: null, slot: null} as any;

    const sigStr = bs58.encode(Buffer.from(rawSig));
    const slotStr = (update.transaction as any)?.slot ?? null;
    return {sig: sigStr, slot: slotStr};
  }

  const statusMap = new Map<string, {slotLS?: string; slotYS?: string}>();

  function maybePrint(sig: string) {
    const entry = statusMap.get(sig);
    if (entry && entry.slotLS && entry.slotYS) {
      console.log(`MATCH ${sig}  LS_slot=${entry.slotLS}  YS_slot=${entry.slotYS}`);
      statusMap.delete(sig);
    }
  }

  // ---------- helper to start a stream ----------
  const startLaserstreamStream = async (
    client: LaserstreamClient,
    onSig: (sig: string) => void,
    label: string
  ): Promise<{ client: any; streamHandle: any }> => {
    // Create client with endpoint and token

    // Subscribe with request and callback for raw protobuf buffer
    const streamHandle = await client.subscribe(subscriptionRequest, (error: Error | null, rawBuffer: Buffer) => {
      if (error) {
        console.error('🚨 LASERSTREAM ERROR:', error.message);
        console.error('   Error type:', error.name);
        console.error('   Full error:', error);
        errLS += 1;
        return;
      }
      
      try {
        // Decode the raw protobuf buffer
        const u = SubscribeUpdate.decode(rawBuffer);

        const {sig, slot} = extractSigAndSlot(u);
        if (!sig || !slot) return;
        onSig(sig);

        slotLS.set(sig, slot);

        const slotNum = Number(slot);
        if (!lsBySlot.has(slotNum)) lsBySlot.set(slotNum, new Set());
        lsBySlot.get(slotNum)!.add(sig);
        if (slotNum > maxSlotLS) maxSlotLS = slotNum;

        const entry: any = statusMap.get(sig) || {};
        entry.slotLS = slot ?? 'unknown';
        statusMap.set(sig, entry);
        maybePrint(sig);
      } catch (decodeError) {
        console.error('Failed to decode Laserstream protobuf:', decodeError);
        errLS += 1;
      }
    });

    // Return both client and stream handle to keep them alive
    return { client, streamHandle };
  };

  // Reference stream using raw Yellowstone gRPC client
  const startYellowstoneStream = async (): Promise<{ client: any; stream: any }> => {
    const client = new Client(yellowstoneConfig.endpoint, yellowstoneConfig.xToken, {
      "grpc.max_receive_message_length": 64 * 1024 * 1024,
    });

    const stream = await client.subscribe();

    // send the request
    await new Promise<void>((resolve, reject) => {
      stream.write(subscriptionRequest, (err: Error | null | undefined) => {
        if (err == null) resolve();
        else reject(err);
      });
    });

    stream.on('data', (u: any) => {
      const {sig, slot} = extractSigAndSlot(u);
      if (!sig || !slot) return;
      gotYS.add(sig);
      newYS += 1;

      slotYS.set(sig, slot);

      const slotNum = Number(slot);
      if (!ysBySlot.has(slotNum)) ysBySlot.set(slotNum, new Set());
      ysBySlot.get(slotNum)!.add(sig);
      if (slotNum > maxSlotYS) maxSlotYS = slotNum;

      const entry: any = statusMap.get(sig) || {};
      entry.slotYS = slot ?? 'unknown';
      statusMap.set(sig, entry);
      maybePrint(sig);
    });

    stream.on('error', (e: Error) => { errYS += 1; console.error('YELLOWSTONE stream error:', e); });
    stream.on('end', () => { errYS += 1; console.error('YELLOWSTONE stream ended'); });

    // Return both client and stream to keep them alive
    return { client, stream };
  };

  // Global variables to keep streams alive
  let laserstreamClient: any = null;
  let laserstreamStream: any = null;
  let yellowstoneClient: any = null;
  let yellowstoneStream: any = null;

  // Start both streams concurrently
  const [laserstreamConnection, yellowstoneConnection] = await Promise.all([
    startLaserstreamStream(client, (sig) => {
      gotLS.add(sig);
      newLS += 1;
    }, 'LASERSTREAM'),
    startYellowstoneStream(),
  ]);

  // Store references globally to prevent garbage collection
  laserstreamClient = laserstreamConnection.client;
  laserstreamStream = laserstreamConnection.streamHandle;
  yellowstoneClient = yellowstoneConnection.client;
  yellowstoneStream = yellowstoneConnection.stream;

  // Cleanup on exit
  process.on('SIGINT', () => {
    if (laserstreamStream && typeof laserstreamStream.cancel === 'function') {
      laserstreamStream.cancel();
    }
    if (yellowstoneStream && typeof yellowstoneStream.end === 'function') {
      yellowstoneStream.end();
    }
    process.exit(0);
  });

  // Keep the script alive indefinitely so we can continue to receive updates.
  await new Promise(() => {});
}

main().catch((err) => {
  console.error('Unhandled error in main:', err);
  process.exit(1);
});

File: go/test/transaction-integrity-test.go
package main

import (
	"context"
	"encoding/hex"
	"fmt"
	"log"
	"math"
	"os"
	"sync"
	"time"

	laserstream "laserstream-go-client"

	"github.com/joho/godotenv"

	pb "github.com/rpcpool/yellowstone-grpc/examples/golang/proto"
	"google.golang.org/grpc"
	"google.golang.org/grpc/credentials/insecure"
	"google.golang.org/grpc/metadata"
)

const (
	PUMP_ADDRESS             = "pAMMBay6oceH9fJKBRHGP5D4bD4sWpmSwMn52FMfXEA"
	SLOT_LAG                 = 3000
	INTEGRITY_CHECK_INTERVAL = 30 * time.Second
)

type SignatureTracker struct {
	mu sync.RWMutex

	gotLS map[string]bool
	gotYS map[string]bool

	slotLS map[string]string
	slotYS map[string]string

	lsBySlot  map[uint64]map[string]bool
	ysBySlot  map[uint64]map[string]bool
	maxSlotLS uint64
	maxSlotYS uint64

	newLS int
	newYS int
	errLS int
	errYS int

	statusMap map[string]map[string]string
}

func NewSignatureTracker() *SignatureTracker {
	return &SignatureTracker{
		gotLS:     make(map[string]bool),
		gotYS:     make(map[string]bool),
		slotLS:    make(map[string]string),
		slotYS:    make(map[string]string),
		lsBySlot:  make(map[uint64]map[string]bool),
		ysBySlot:  make(map[uint64]map[string]bool),
		statusMap: make(map[string]map[string]string),
	}
}

func (st *SignatureTracker) addLaserstream(sig string, slot uint64) {
	st.mu.Lock()
	defer st.mu.Unlock()

	slotStr := fmt.Sprintf("%d", slot)

	if !st.gotLS[sig] {
		st.gotLS[sig] = true
		st.newLS++
	}

	st.slotLS[sig] = slotStr

	if st.lsBySlot[slot] == nil {
		st.lsBySlot[slot] = make(map[string]bool)
	}
	st.lsBySlot[slot][sig] = true

	if slot > st.maxSlotLS {
		st.maxSlotLS = slot
	}

	if st.statusMap[sig] == nil {
		st.statusMap[sig] = make(map[string]string)
	}
	st.statusMap[sig]["slotLS"] = slotStr
	st.maybePrint(sig)
}

func (st *SignatureTracker) addYellowstone(sig string, slot uint64) {
	st.mu.Lock()
	defer st.mu.Unlock()

	slotStr := fmt.Sprintf("%d", slot)

	if !st.gotYS[sig] {
		st.gotYS[sig] = true
		st.newYS++
	}

	st.slotYS[sig] = slotStr

	if st.ysBySlot[slot] == nil {
		st.ysBySlot[slot] = make(map[string]bool)
	}
	st.ysBySlot[slot][sig] = true

	if slot > st.maxSlotYS {
		st.maxSlotYS = slot
	}

	if st.statusMap[sig] == nil {
		st.statusMap[sig] = make(map[string]string)
	}
	st.statusMap[sig]["slotYS"] = slotStr
	st.maybePrint(sig)
}

func (st *SignatureTracker) maybePrint(sig string) {
	entry := st.statusMap[sig]
	if entry != nil && entry["slotLS"] != "" && entry["slotYS"] != "" {
		log.Printf("MATCH %s LS_slot=%s YS_slot=%s", sig, entry["slotLS"], entry["slotYS"])
		delete(st.statusMap, sig)
	}
}

func (st *SignatureTracker) addError(isLS bool) {
	st.mu.Lock()
	defer st.mu.Unlock()

	if isLS {
		st.errLS++
	} else {
		st.errYS++
	}
}

func (st *SignatureTracker) doIntegrityCheck() {
	st.mu.Lock()
	defer st.mu.Unlock()

	readySlot := uint64(math.Min(float64(st.maxSlotLS), float64(st.maxSlotYS))) - SLOT_LAG

	slotsToProcess := make(map[uint64]bool)
	for slot := range st.lsBySlot {
		if slot <= readySlot {
			slotsToProcess[slot] = true
		}
	}
	for slot := range st.ysBySlot {
		if slot <= readySlot {
			slotsToProcess[slot] = true
		}
	}

	totalMissingLS := 0
	totalMissingYS := 0

	for slot := range slotsToProcess {
		setLS := st.lsBySlot[slot]
		setYS := st.ysBySlot[slot]
		if setLS == nil {
			setLS = make(map[string]bool)
		}
		if setYS == nil {
			setYS = make(map[string]bool)
		}

		var missingLS []string
		var missingYS []string

		for sig := range setYS {
			if !setLS[sig] {
				missingLS = append(missingLS, sig)
			}
		}
		for sig := range setLS {
			if !setYS[sig] {
				missingYS = append(missingYS, sig)
			}
		}

		if len(missingLS) > 0 || len(missingYS) > 0 {
			log.Printf("[INTEGRITY] transaction_mismatch slot=%d missing Laserstream=%d missing Yellowstone=%d",
				slot, len(missingLS), len(missingYS))
			for _, sig := range missingLS {
				log.Printf("SIGNATURE MISSING IN LASERSTREAM %s (YS_slot=%d)", sig, slot)
			}
			for _, sig := range missingYS {
				log.Printf("SIGNATURE MISSING IN YELLOWSTONE %s (LS_slot=%d)", sig, slot)
			}
		}

		totalMissingLS += len(missingLS)
		totalMissingYS += len(missingYS)

		delete(st.lsBySlot, slot)
		delete(st.ysBySlot, slot)
	}

	now := time.Now().Format(time.RFC3339)
	log.Printf("[%s] laserstream+%d yellowstone+%d processedSlots:%d missingLS:%d missingYS:%d LS_errors:%d YS_errors:%d",
		now, st.newLS, st.newYS, len(slotsToProcess), totalMissingLS, totalMissingYS, st.errLS, st.errYS)

	st.newLS = 0
	st.newYS = 0
}

func extractSigAndSlot(update *laserstream.SubscribeUpdate) (string, uint64, bool) {
	if txUpdate, ok := update.UpdateOneof.(*laserstream.SubscribeUpdate_Transaction); ok {
		if txUpdate.Transaction != nil && txUpdate.Transaction.Transaction != nil {
			if len(txUpdate.Transaction.Transaction.Signature) > 0 {
				sig := hex.EncodeToString(txUpdate.Transaction.Transaction.Signature)
				return sig, txUpdate.Transaction.Slot, true
			}
		}
	}
	return "", 0, false
}

func extractSigAndSlotYS(resp *pb.SubscribeUpdate) (string, uint64, bool) {
	if txUpdate, ok := resp.UpdateOneof.(*pb.SubscribeUpdate_Transaction); ok {
		if txUpdate.Transaction != nil && txUpdate.Transaction.Transaction != nil {
			if len(txUpdate.Transaction.Transaction.Signature) > 0 {
				sig := hex.EncodeToString(txUpdate.Transaction.Transaction.Signature)
				return sig, txUpdate.Transaction.Slot, true
			}
		}
	}
	return "", 0, false
}

func startLaserstreamStream(tracker *SignatureTracker) error {
	endpoint := os.Getenv("LASERSTREAM_ENDPOINT")
	if endpoint == "" {
		endpoint = "localhost:4003" // Default to chaos proxy
	}
	apiKey := os.Getenv("LASERSTREAM_API_KEY")
	if apiKey == "" {
		log.Fatal("LASERSTREAM_API_KEY required")
	}

	config := laserstream.LaserstreamConfig{
		Endpoint: endpoint,
		APIKey:   apiKey,
		Insecure: true, // localhost connection through chaos proxy
	}

	commitmentLevel := laserstream.CommitmentLevel_CONFIRMED
	voteFilterBool := false
	failedFilterBool := false

	subscriptionRequest := &laserstream.SubscribeRequest{
		Transactions: map[string]*laserstream.SubscribeRequestFilterTransactions{
			"client": {
				AccountInclude: []string{PUMP_ADDRESS},
				Vote:           &voteFilterBool,
				Failed:         &failedFilterBool,
			},
		},
		Commitment: &commitmentLevel,
	}

	client := laserstream.NewClient(config)

	dataCallback := func(data *laserstream.SubscribeUpdate) {
		if sig, slot, ok := extractSigAndSlot(data); ok {
			tracker.addLaserstream(sig, slot)
		}
	}

	errorCallback := func(err error) {
		tracker.addError(true)
		log.Printf("LASERSTREAM error: %v", err)
	}

	return client.Subscribe(subscriptionRequest, dataCallback, errorCallback)
}

func startYellowstoneStream(tracker *SignatureTracker) error {
	endpoint := os.Getenv("YELLOWSTONE_ENDPOINT")
	if endpoint == "" {
		log.Fatal("YELLOWSTONE_ENDPOINT required")
	}
	apiKey := os.Getenv("YELLOWSTONE_API_KEY")
	if apiKey == "" {
		log.Fatal("YELLOWSTONE_API_KEY required")
	}

	conn, err := grpc.Dial(endpoint, grpc.WithTransportCredentials(insecure.NewCredentials()))
	if err != nil {
		return fmt.Errorf("failed to connect to yellowstone: %w", err)
	}

	client := pb.NewGeyserClient(conn)

	ctx := context.Background()
	md := metadata.New(map[string]string{"x-token": apiKey})
	ctx = metadata.NewOutgoingContext(ctx, md)

	stream, err := client.Subscribe(ctx)
	if err != nil {
		return fmt.Errorf("failed to create yellowstone stream: %w", err)
	}

	pbCommitmentLevel := pb.CommitmentLevel_CONFIRMED
	voteFilterBool := false
	failedFilterBool := false

	subscriptionRequest := &pb.SubscribeRequest{
		Transactions: map[string]*pb.SubscribeRequestFilterTransactions{
			"client": {
				AccountInclude: []string{PUMP_ADDRESS},
				Vote:           &voteFilterBool,
				Failed:         &failedFilterBool,
			},
		},
		Commitment: &pbCommitmentLevel,
	}

	if err := stream.Send(subscriptionRequest); err != nil {
		return fmt.Errorf("failed to send yellowstone subscription: %w", err)
	}

	go func() {
		for {
			resp, err := stream.Recv()
			if err != nil {
				tracker.addError(false)
				log.Printf("YELLOWSTONE stream error: %v", err)
				return
			}

			if sig, slot, ok := extractSigAndSlotYS(resp); ok {
				tracker.addYellowstone(sig, slot)
			}
		}
	}()

	return nil
}

func main() {
	log.SetFlags(0)

	godotenv.Load("../.env")

	tracker := NewSignatureTracker()

	go func() {
		ticker := time.NewTicker(INTEGRITY_CHECK_INTERVAL)
		defer ticker.Stop()
		for range ticker.C {
			tracker.doIntegrityCheck()
		}
	}()

	go func() {
		if err := startLaserstreamStream(tracker); err != nil {
			log.Fatalf("Failed to start Laserstream: %v", err)
		}
	}()

	go func() {
		if err := startYellowstoneStream(tracker); err != nil {
			log.Fatalf("Failed to start Yellowstone: %v", err)
		}
	}()

	log.Println("Both streams started. Running integrity test...")

	select {}
}
